src/middleware.coffee

```diff
@@ -21,32 +22,44 @@ class Middleware
   # done() - Initial (final) completion callback. May be wrapped by
   #     executed middleware.
   #
-  # Returns nothing
+  # Returns promise - resolves with context when middleware completes
   # Returns before executing any middleware
   execute: (context, next, done) ->
-    done ?= ->
-    # Execute a single piece of middleware and update the completion callback
-    # (each piece of middleware can wrap the 'done' callback with additional
-    # logic).
-    executeSingleMiddleware = (doneFunc, middlewareFunc, cb) =>
-      # Match the async.reduce interface
-      nextFunc = (newDoneFunc) -> cb(null, newDoneFunc or doneFunc)
-      # Catch errors in synchronous middleware
-      try
-        middlewareFunc.call(undefined, context, nextFunc, doneFunc)
-      catch err
-        # Maintaining the existing error interface (Response object)
-        @robot.emit('error', err, context.response)
-        # Forcibly fail the middleware and stop executing deeper
-        doneFunc()
 
-    # Executed when the middleware stack is finished
-    allDone = (_, finalDoneFunc) -> next(context, finalDoneFunc)
+    new Promise (resolve, reject) =>
 
-    # Execute each piece of middleware, collecting the latest 'done' callback
-    # at each step.
-    process.nextTick =>
-      async.reduce(@stack, done, executeSingleMiddleware, allDone)
+      done ?= ->
+        
+      # Allow each middleware to resolve the promise early if it calls done()
+      pieceDone = ->
+        done()
+        resolve context
+        
+      # Execute a single piece of middleware and update the completion callback
+      # (each piece of middleware can wrap the 'done' callback with additional
+      # logic).
+      executeSingleMiddleware = (doneFunc, middlewareFunc, cb) =>
+        # Match the async.reduce interface
+        nextFunc = (newDoneFunc) -> cb(null, newDoneFunc or doneFunc)
+        # Catch errors in synchronous middleware
+        try
+          middlewareFunc.call(undefined, context, nextFunc, doneFunc)
+        catch err
+          # Maintaining the existing error interface (Response object)
+          @robot.emit('error', err, context.response)
+          # Forcibly fail the middleware and stop executing deeper
+          doneFunc()
+          reject err, context
```

Comment:
`Promise.reject()` only accepts one argument, you can try it out using

```js
Promise.reject(new Error('foo'), {}).catch(function (error, context) {
  console.log(context)
})
// logs `undefined`
```

What you could do is set context as property of `err` before rejecting with it